cmd/sharded-test-server/frontproxy.go

```diff
@@ -53,18 +61,39 @@ func startFrontProxy(ctx context.Context, args []string, servingCA *crypto.CA, h
 		lineprefix.Color(color.New(color.FgHiWhite)),
 	)
 
-	if err := ioutil.WriteFile(filepath.Join(workDirPath, ".kcp-front-proxy/mapping.yaml"), []byte(`
-- path: /services/
-  backend: https://localhost:6444
-  backend_server_ca: .kcp/serving-ca.crt
-  proxy_client_cert: .kcp-front-proxy/requestheader.crt
-  proxy_client_key: .kcp-front-proxy/requestheader.key
-- path: /clusters/
-  backend: https://localhost:6444
-  backend_server_ca: .kcp/serving-ca.crt
-  proxy_client_cert: .kcp-front-proxy/requestheader.crt
-  proxy_client_key: .kcp-front-proxy/requestheader.key
-`), 0644); err != nil {
+	type mappingEntry struct {
+		Path            string `json:"path"`
+		Backend         string `json:"backend"`
+		BackendServerCA string `json:"backend_server_ca"`
+		ProxyClientCert string `json:"proxy_client_cert"`
+		ProxyClientKey  string `json:"proxy_client_key"`
+	}
+
+	mappings := []mappingEntry{
+		{
+			Path: "/services/",
+			// TODO: support multiple virtual workspace backend servers
+			Backend:         fmt.Sprintf("https://localhost:%s", vwPort),
+			BackendServerCA: ".kcp/serving-ca.crt",
+			ProxyClientCert: ".kcp-front-proxy/requestheader.crt",
+			ProxyClientKey:  ".kcp-front-proxy/requestheader.key",
+		},
+		{
+			Path: "/clusters/",
+			// TODO: support multiple shard backend servers
+			Backend:         "https://localhost:6444",
+			BackendServerCA: ".kcp/serving-ca.crt",
+			ProxyClientCert: ".kcp-front-proxy/requestheader.crt",
+			ProxyClientKey:  ".kcp-front-proxy/requestheader.key",
+		},
+	}
```

Comment:
for follow-up: we should really turn this mapping file into flags. The file promises flexibility that is not there.

Reply:
Now there is: https://github.com/kcp-dev/kcp/issues/1857